benchmark/Streamly/Benchmark/Data/Stream/StreamK.hs

```diff
@@ -195,125 +167,173 @@ composeN n f =
 
--- this is quadratic
-iterateScan            = iterateSource (S.scanl' (+) 0) (maxIters `div` 10)
-iterateDropWhileFalse  = iterateSource (S.dropWhile (> maxValue))
-                                       (maxIters `div` 10)
+{-# INLINE iterateDropOne #-}
+iterateDropOne :: S.MonadAsync m => Int -> Int -> Int -> Stream m Int
+iterateDropOne iterStreamLen maxIters = iterateSource iterStreamLen (S.drop 1) maxIters
 
-iterateMapM            = iterateSource (S.mapM return) maxIters
-iterateFilterEven      = iterateSource (S.filter even) maxIters
-iterateTakeAll         = iterateSource (S.take maxValue) maxIters
-iterateDropOne         = iterateSource (S.drop 1) maxIters
-iterateDropWhileTrue   = iterateSource (S.dropWhile (<= maxValue)) maxIters
+{-# INLINE iterateDropWhileTrue #-}
+iterateDropWhileTrue :: S.MonadAsync m => Int -> Int -> Int -> Int -> Stream m Int
+iterateDropWhileTrue streamLen iterStreamLen maxIters = iterateSource iterStreamLen (S.dropWhile (<= streamLen)) maxIters
 
 -------------------------------------------------------------------------------
 -- Zipping
 -------------------------------------------------------------------------------
 
-zip src       = transform $ S.zipWith (,) src src
+{-# INLINE zip #-}
+zip :: Monad m => Stream m Int -> m ()
+zip src = transform $ S.zipWith (,) src src
 
 -------------------------------------------------------------------------------
 -- Mixed Composition
 -------------------------------------------------------------------------------
 
 {-# INLINE scanMap #-}
+scanMap :: Monad m => Int -> Stream m Int -> m ()
+scanMap n = composeN n $ S.map (subtract 1) . S.scanl' (+) 0
+
 {-# INLINE dropMap #-}
+dropMap :: Monad m => Int -> Stream m Int -> m ()
+dropMap n = composeN n $ S.map (subtract 1) . S.drop 1
+
 {-# INLINE dropScan #-}
+dropScan :: Monad m => Int -> Stream m Int -> m ()
+dropScan n = composeN n $ S.scanl' (+) 0 . S.drop 1
+
 {-# INLINE takeDrop #-}
+takeDrop :: Monad m => Int -> Int -> Stream m Int -> m ()
+takeDrop streamLen n = composeN n $ S.drop 1 . S.take streamLen
+
 {-# INLINE takeScan #-}
+takeScan :: Monad m => Int -> Int -> Stream m Int -> m ()
+takeScan streamLen n = composeN n $ S.scanl' (+) 0 . S.take streamLen
+
 {-# INLINE takeMap #-}
+takeMap :: Monad m => Int -> Int -> Stream m Int -> m ()
+takeMap streamLen n = composeN n $ S.map (subtract 1) . S.take streamLen
+
 {-# INLINE filterDrop #-}
+filterDrop :: Monad m => Int -> Int -> Stream m Int -> m ()
+filterDrop streamLen n = composeN n $ S.drop 1 . S.filter (<= streamLen)
+
 {-# INLINE filterTake #-}
+filterTake :: Monad m => Int -> Int -> Stream m Int -> m ()
+filterTake streamLen n = composeN n $ S.take streamLen . S.filter (<= streamLen)
+
+-- Initially, this used maxBound not streamLen
 {-# INLINE filterScan #-}
+filterScan :: Monad m => Int -> Int -> Stream m Int -> m ()
+filterScan streamLen n = composeN n $ S.scanl' (+) 0 . S.filter (<= streamLen)
```

Comment:
Since we are adding the numbers in the stream, `<= streamLen` would terminate the stream much earlier. We should use `maxBound` instead.